src/binding/c/pt2pt_api.txt

```diff
@@ -0,0 +1,557 @@
+# vim: set ft=c:
+
+MPI_Bsend:
+    .file: pt2pt/bsend.c
+    .desc: Basic send with user-provided buffering
+    .seealso: MPI_Buffer_attach, MPI_Ibsend, MPI_Bsend_init
+    .earlyreturn: pt2pt_proc_null
+/*
+    Notes:
+    This send is provided as a convenience function; it allows the user to
+    send messages without worring about where they are buffered (because the
+    user `must` have provided buffer space with 'MPI_Buffer_attach').
+
+    In deciding how much buffer space to allocate, remember that the buffer space
+    is not available for reuse by subsequent 'MPI_Bsend's unless you are certain
+    that the message
+    has been received (not just that it should have been received).  For example,
+    this code does not allocate enough buffer space
+    .vb
+        MPI_Buffer_attach(b, n*sizeof(double) + MPI_BSEND_OVERHEAD);
+        for (i=0; i<m; i++) {
+            MPI_Bsend(buf, n, MPI_DOUBLE, ...);
+        }
+    .ve
+    because only enough buffer space is provided for a single send, and the
+    loop may start a second 'MPI_Bsend' before the first is done making use of the
+    buffer.
+
+    In C, you can
+    force the messages to be delivered by
+    .vb
+        MPI_Buffer_detach(&b, &n);
+        MPI_Buffer_attach(b, n);
+    .ve
+    (The 'MPI_Buffer_detach' will not complete until all buffered messages are
+    delivered.)
+*/
+{
+    mpi_errno = MPIR_Bsend_isend(buf, count, datatype, dest, tag, comm_ptr, NULL);
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+}
```

Comment:
I don't understand the benefit here. Isn't this code that should be autogenerated too?

My understanding was that we'd autogenerate all of the error checking and then call the `MPIR` functions for everything else.

Reply:
You are questioning the reason to allow direct code in the configuration file? It is for flexibility, so we have a back-door to customize certain functions when it is too complex or awkward to implement in the script or support new configure options.

In the case of `MPI_Bsend`, it is calling `MPIR_Bsend_isend` in stead of `MPIR_Bsend_impl` and having to wrap it with another `MPIR_Bsend_impl` seems too awkward :)

In addtion, the MPIR-impl functions are not inlined so for especially the pt2pt path, having direct code calling `MPID-` code may be desirable.

Reply:
> Are you using something like f-strings? I think we use that in the MPI Standard repo and that's pretty new (3.8 I think?). If this is only a requirement for the developers, that's probably not a big deal, but if it's something that gets run at configure time, that might be harder.

No, I am using mostly '%' string -- no advanced features. The idea to require Python 3 is so we don't have to deal with any `Python 2` issues and since Python 2 is deprecated, all new scripts should target Python 3 anyway.

The scripts are only used during `autogen`, not `configure`, so I guess it is OK to be more picky in autogen. Not this, but we do have scripts in configure command, so I hope requiring Python 3 is not deal-breaker even there. (Here is another thing that reminds how great Perl is :) ).

Reply:
I agree with @wesbland that some of the code here is probably better suited to autogeneration. But note that all the user argument checking **is** autogenerated, which is a big win IMO.

I am not opposed to the concept of the config file. We have this type of capability in `buildiface` scripts, too. Perhaps more aggressive generation doesn't need to happen in this first pass.